src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.ReplicationLoader.cs

```diff
@@ -130,14 +130,34 @@ public void InvokeOnFailed(ShardedIncomingReplicationHandler handler, Exception
         }
     }
 
-    public class ReplicationBatch
+    public class ReplicationBatch : IDisposable
     {
         public List<ReplicationBatchItem> Items = new();
-        public Dictionary<Slice, AttachmentReplicationItem> Attachments;
+        public Dictionary<Slice, AttachmentReplicationItem> AttachmentStreams;
         public TaskCompletionSource BatchSent;
         public string LastAcceptedChangeVector;
         public long LastEtagAccepted;
         public long LastSentEtagFromSource;
+
+        public void Dispose()
+        {
+            if (AttachmentStreams != null)
+            {
+                foreach (var attachment in AttachmentStreams)
+                    attachment.Value?.Dispose();
+
+                AttachmentStreams.Clear();
+            }
+
+            foreach (var item in Items)
+            {
+                item.Dispose();
+            }
+
+            Items.Clear();
+            BatchSent?.TrySetCanceled();
```

Comment:
we are cancelling if after we are clearing and disposing. Is that OK? We are not accessing those items anywhere else at the same time?

Reply:
We should do it at the beginning, done